src/Swift.Bindings/src/Emitter/StringCSharpEmitter/Handler/MethodHandler.cs

```diff
@@ -58,167 +150,243 @@ public IEnvironment Marshal(BaseDecl methodDecl)
+        /// <param name="methodDecl">The method declaration.</param>
+        /// <param name="parentDecl">The parent declaration.</param>
+        /// <param name="typeDatabase">The type database.</param>
+        public PInvokeSignatureBuilder(MethodDecl methodDecl, BaseDecl parentDecl, TypeDatabase typeDatabase)
         {
-            var methodDecl = (MethodDecl)env.MethodDecl;
-            var parentDecl = methodDecl.ParentDecl ?? throw new ArgumentNullException(nameof(methodDecl.ParentDecl));
-            var moduleDecl = methodDecl.ModuleDecl ?? throw new ArgumentNullException(nameof(methodDecl.ParentDecl));
-
-            string methodName = $"{env.PInvokePrefix}{methodDecl.Name}";
-
-            writer.WriteLine($"public {(methodDecl.MethodType == MethodType.Static ? "static " : "")}{methodDecl.CSSignature.First().CSTypeIdentifier.Name} {methodDecl.Name}({GetPublicMethodSignature(methodDecl)})");
-            writer.WriteLine("{");
-            writer.Indent++;
+            MethodDecl = methodDecl;
+            ParentDecl = parentDecl;
+            TypeDatabase = typeDatabase;
+        }
 
-            if (methodDecl.MethodType == MethodType.Instance)
+        /// <summary>
+        /// Handles the return type of the method.
+        /// </summary>
+        public void HandleReturnType()
+        {
+            if (!MethodDecl.RequiresIndirectResult(ParentDecl, TypeDatabase))
             {
-                writer.WriteLine($"{parentDecl.Name} self = this;");
+                var returnType = MethodDecl.CSSignature.First().CSTypeIdentifier.Name;
+                SetReturnType(returnType);
+            }
+            else
+            {
+                AddParameter("SwiftIndirectResult", "swiftIndirectResult");
+                SetReturnType("void");
             }
-            string returnPrefix = methodDecl.CSSignature.First().CSTypeIdentifier.Name == "void" ? "" : "return ";
-            string methodArgs = string.Join(", ", methodDecl.CSSignature.Skip(1).Select(p => p.Name));
-            writer.WriteLine($"{returnPrefix}{methodName}({GetMethodArgs(methodDecl)});");
-            writer.Indent--;
-            writer.WriteLine("}");
         }
 
         /// <summary>
-        /// Gets the method parameters.
+        /// Handles the arguments of the method.
         /// </summary>
-        /// <param name="methodDecl">The method declaration.</param>
-        /// <returns>The list of method parameters.</returns>
-        private List<ArgumentDecl> GetMethodParams(MethodDecl methodDecl)
+        public void HandleArguments()
         {
-            var parentDecl = methodDecl.ParentDecl ?? throw new ArgumentNullException(nameof(methodDecl.ParentDecl));
-            List<ArgumentDecl> tempDecl = new(methodDecl.CSSignature);
+            foreach (var argument in MethodDecl.CSSignature.Skip(1))
+            {
+                var typeRecord = TypeDatabase.Registrar.GetType(argument);
```

Comment:
This should be a TryGet pattern since the type may not exist. In addition, if the type database entry comes from an external module, you will either need to use the full type name of inject a using statement.

Reply:
We should always use fully qualified names - to avoid unintentional collisions. In fact we should use the `global:` prefix as well to make sure we don't pick up some other random namespace item. Just like Roslyn source generator.

Reply:
@stephen-hawley, I'm thinking about the behaviour on TryGet method failure.  Should we just skip this methods projection? Or we should risk getting it wrong?

Reply:
In BTfS for this class of failure it was treated as a warning and a skip for the method.

Reply:
We could do this by creating something like "transactions" - when we're about to start projecting a method, we "start a transaction" which will reset all output buffers. Then we project/write and everything goes into output buffers, only when we're successfully done, we "commit" and appends the buffers to the actual output.
This would be more robust than trying to check for every error condition upfront - it's easy to miss something, and that doesn't count actual unexpected exceptions during the writing process (our bugs and so on). It would be really nice if the tool was resilient to its own bugs to some degree.

Reply:
> if the type database entry comes from an external module, you will either need to use the full type name of inject a using statement.

I agree. I think we should add some mechanism which allows us to resolve a fully qualified c# name from any `Decl` and any TypeDatabase entry. This should probably be bundled with the translation of unpronouncable names is csharp. So lets track it together. Out of scope of this PR. https://github.com/dotnet/runtimelab/issues/2889

>This should be a TryGet pattern since the type may not exist. 

I like the "transactional" idea. Again I think it will require some separate discussion. Lets do it separately. Tracking issue: https://github.com/dotnet/runtimelab/issues/2890.